compiler/next/include/chpl/AST/UniqueStringDetail.h

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  This file implements a global table of strings which enables easier string
+  comparison and memory management.
+ */
+
+#ifndef CHPL_AST_UNIQUE_STRING_DETAIL_H
+#define CHPL_AST_UNIQUE_STRING_DETAIL_H
+
+#include <cassert>
+#include <cstring>
+#include <string>
+
+/// \cond DO_NOT_DOCUMENT
+namespace chpl {
+class Context;
+namespace ast {
+namespace detail {
+
+// We can make it store 6 bytes in line this way:
+// alloc all such strings aligned to 2 bytes
+// store in the low-order byte an odd number
+// store in the last (memory order) byte, 6-length
+//   (so we store 0 if the length is 6, which is the terminator)a
+// fill any unused bytes with 0s
+//
+// This should work on both big and little endian systems.
+// Q: what is the string distribution in Chapel programs?
+//   compiling distributions primer with something like 1.24 we see:
+//
+//            ~1,075,000 table queries            vs English text (moby dick)
+//   of these,  ~526,000 (49%) are < 7 bytes      (78%)
+//              ~875,000 (81%) are < 13 bytes     (98%)
+//            ~1,008,000 (94%) are < 23 bytes
+//            ~1,034,000 (96%) are < 30 bytes
+//            ~1,050,000 (98%) are < 39 bytes
+//
+//            ~45,000 unique strings
+//   of these,  ~8,000 (18%) are < 7 bytes
+//             ~12,000 (25%) are < 13 bytes
+//             ~24,000 (53%) are < 23 bytes
+//             ~28,000 (62%) are < 30 bytes
+//             ~33,000 (73%) are < 39 bytes
+//
+// Q: does this improve performance?
+//   Yes, by about 5-10% in testUniqueString.cpp
+//   It might be more than that when the string references are
+//   not recently used.
+
+struct InlinedString {
+
+  // This union is only used to make the debugger output more reasonable.
+  // All reads/writes here go through the 'v' variable.
+  // Performance: Should we "inline" strings up to 14 bytes
+  // by using a 16 byte data region? To try that, some of the
+  // surrounding code will need adjustment.
+  union {
+    const char* v;
+    char data[sizeof(const char*)];
+  };
+
```

Comment:
BTW, this is what is meant in the C/C++ literature as an "anonymous union". All of the members of the anonymous union can be accessed as members of the enclosing struct/class/union.

"Untagged Union" would be a more appropriate name for a union which does not have a tag (after the `union` keyword).